File: src/api/types/rules.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize)]
pub struct RulesListResponse {
    pub(crate) rules: Vec<RuleResponse>,
}

#[derive(Debug, Serialize)]
pub struct RuleResponse {
    pub(crate) id: String,
    pub(crate) name: String,
    pub(crate) enabled: bool,
    pub(crate) layer: String,
    pub(crate) pattern: String,
    pub(crate) action: String,
    pub(crate) severity: String,
    pub(crate) plugin_template_id: Option<String>,
    pub(crate) response_template: Option<RuleResponseTemplatePayload>,
}

#[derive(Debug, Deserialize)]
pub struct RuleUpsertRequest {
    pub(crate) id: String,
    pub(crate) name: String,
    pub(crate) enabled: bool,
    pub(crate) layer: String,
    pub(crate) pattern: String,
    pub(crate) action: String,
    pub(crate) severity: String,
    pub(crate) plugin_template_id: Option<String>,
    pub(crate) response_template: Option<RuleResponseTemplatePayload>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuleResponseTemplatePayload {
    pub(crate) status_code: u16,
    pub(crate) content_type: String,
    pub(crate) body_source: String,
    pub(crate) gzip: bool,
    pub(crate) body_text: String,
    pub(crate) body_file_path: String,
    #[serde(default)]
    pub(crate) headers: Vec<RuleResponseHeaderPayload>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuleResponseHeaderPayload {
    pub(crate) key: String,
    pub(crate) value: String,
}

#[derive(Debug, Serialize)]
pub struct RuleActionPluginsResponse {
    pub(crate) total: u32,
    pub(crate) plugins: Vec<RuleActionPluginResponse>,
}

#[derive(Debug, Serialize)]
pub struct RuleActionPluginResponse {
    pub(crate) plugin_id: String,
    pub(crate) name: String,
    pub(crate) version: String,
    pub(crate) description: String,
    pub(crate) enabled: bool,
    pub(crate) installed_at: i64,
    pub(crate) updated_at: i64,
}

#[derive(Debug, Serialize)]
pub struct RuleActionTemplatesResponse {
    pub(crate) total: u32,
    pub(crate) templates: Vec<RuleActionTemplateResponse>,
}

#[derive(Debug, Serialize)]
pub struct RuleActionTemplateResponse {
    pub(crate) template_id: String,
    pub(crate) plugin_id: String,
    pub(crate) name: String,
    pub(crate) description: String,
    pub(crate) layer: String,
    pub(crate) action: String,
    pub(crate) pattern: String,
    pub(crate) severity: String,
    pub(crate) response_template: RuleResponseTemplatePayload,
    pub(crate) updated_at: i64,
}

#[derive(Debug, Serialize)]
pub struct RuleActionTemplatePreviewResponse {
    pub(crate) template_id: String,
    pub(crate) name: String,
    pub(crate) content_type: String,
    pub(crate) status_code: u16,
    pub(crate) gzip: bool,
    pub(crate) body_source: String,
    pub(crate) body_preview: String,
    pub(crate) truncated: bool,
}

#[derive(Debug, Serialize)]
pub struct ActionIdeaPresetsResponse {
    pub(crate) total: u32,
    pub(crate) ideas: Vec<ActionIdeaPresetResponse>,
}

#[derive(Debug, Serialize)]
pub struct ActionIdeaPresetResponse {
    pub(crate) id: String,
    pub(crate) title: String,
    pub(crate) mood: String,
    pub(crate) summary: String,
    pub(crate) mechanism: String,
    pub(crate) performance: String,
    pub(crate) fallback_path: String,
    pub(crate) plugin_id: String,
    pub(crate) file_name: String,
    pub(crate) response_file_path: String,
    pub(crate) plugin_name: String,
    pub(crate) plugin_description: String,
    pub(crate) template_local_id: String,
    pub(crate) template_name: String,
    pub(crate) template_description: String,
    pub(crate) pattern: String,
    pub(crate) severity: String,
    pub(crate) content_type: String,
    pub(crate) status_code: u16,
    pub(crate) gzip: bool,
    pub(crate) body_source: String,
    pub(crate) runtime_body_file_path: String,
    pub(crate) headers: Vec<RuleResponseHeaderPayload>,
    pub(crate) response_content: String,
    pub(crate) requires_upload: bool,
    pub(crate) uploaded_file_name: Option<String>,
    pub(crate) uploaded_file_ready: bool,
    pub(crate) uploaded_body_preview: Option<String>,
    pub(crate) uploaded_body_preview_notice: Option<String>,
    pub(crate) uploaded_body_truncated: bool,
    pub(crate) has_overrides: bool,
    pub(crate) updated_at: i64,
}

#[derive(Debug, Deserialize)]
pub struct UpdateActionIdeaPresetRequest {
    pub(crate) title: String,
    pub(crate) status_code: u16,
    pub(crate) content_type: String,
    pub(crate) response_content: String,
}

#[derive(Debug, Serialize)]
pub struct ActionIdeaUploadResponse {
    pub(crate) idea: ActionIdeaPresetResponse,
}

#[derive(Debug, Deserialize)]
pub struct InstallRuleActionPluginRequest {
    pub(crate) package_url: String,
    #[serde(default)]
    pub(crate) sha256: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateRuleActionPluginRequest {
    pub(crate) enabled: bool,
}

#[derive(Debug, Serialize)]
pub struct WriteStatusResponse {
    pub(crate) success: bool,
    pub(crate) message: String,
}

File: dome-vue/src/features/rules/types/rules.ts
export interface RuleItem {
  id: string
  name: string
  enabled: boolean
  layer: string
  pattern: string
  action: string
  severity: string
  plugin_template_id?: string | null
  response_template?: RuleResponseTemplate | null
}

export interface RuleDraft {
  id: string
  name: string
  enabled: boolean
  layer: string
  pattern: string
  action: string
  severity: string
  plugin_template_id?: string | null
  response_template?: RuleResponseTemplate | null
}

export interface RuleResponseHeader {
  key: string
  value: string
}

export interface RuleResponseTemplate {
  status_code: number
  content_type: string
  body_source: 'inline_text' | 'file' | string
  gzip: boolean
  body_text: string
  body_file_path: string
  headers: RuleResponseHeader[]
}

export interface RuleActionPluginItem {
  plugin_id: string
  name: string
  version: string
  description: string
  enabled: boolean
  installed_at: number
  updated_at: number
}

export interface RuleActionPluginsResponse {
  total: number
  plugins: RuleActionPluginItem[]
}

export interface RuleActionTemplateItem {
  template_id: string
  plugin_id: string
  name: string
  description: string
  layer: string
  action: string
  pattern: string
  severity: string
  response_template: RuleResponseTemplate
  updated_at: number
}

export interface RuleActionTemplatesResponse {
  total: number
  templates: RuleActionTemplateItem[]
}

export interface RuleActionTemplatePreviewResponse {
  template_id: string
  name: string
  content_type: string
  status_code: number
  gzip: boolean
  body_source: string
  body_preview: string
  truncated: boolean
}

export interface ActionIdeaPreset {
  id: string
  title: string
  mood: string
  summary: string
  mechanism: string
  performance: string
  fallback_path: string
  plugin_id: string
  file_name: string
  response_file_path: string
  plugin_name: string
  plugin_description: string
  template_local_id: string
  template_name: string
  template_description: string
  pattern: string
  severity: string
  content_type: string
  status_code: number
  gzip: boolean
  body_source: string
  runtime_body_file_path: string
  headers: RuleResponseHeader[]
  response_content: string
  requires_upload: boolean
  uploaded_file_name?: string | null
  uploaded_file_ready: boolean
  uploaded_body_preview?: string | null
  uploaded_body_preview_notice?: string | null
  uploaded_body_truncated: boolean
  has_overrides: boolean
  updated_at: number
}

export interface ActionIdeaPresetsResponse {
  total: number
  ideas: ActionIdeaPreset[]
}

export interface RulesResponse {
  rules: RuleItem[]
}
